zivid_camera/test/test_zivid_camera.cpp

```diff
@@ -78,133 +106,276 @@ class ZividNodeTestBase : public testing::Test
+  static constexpr auto points_xyzrgba_topic_name = "points/xyzrgba";
+  static constexpr auto normals_xyz_topic_name = "normals/xyz";
   static constexpr auto file_camera_path = "/usr/share/Zivid/data/FileCameraZivid2M70.zfc";
 
+  static constexpr auto parameter_settings_file_path = "settings_file_path";
+  static constexpr auto parameter_settings_yaml = "settings_yaml";
+  static constexpr auto parameter_settings_2d_file_path = "settings_2d_file_path";
+  static constexpr auto parameter_settings_2d_yaml = "settings_2d_yaml";
+
+  ZividNodeTest() : test_node_(rclcpp::Node::make_shared("test_node"))
+  {
+    executor_.add_node(test_node_);
+    executor_.add_node(zivid_ros_node);
+
+    // Reset test state
+    setNodeParameter(parameter_settings_file_path, "");
+    setNodeParameter(parameter_settings_yaml, "");
+    setNodeParameter(parameter_settings_2d_file_path, "");
+    setNodeParameter(parameter_settings_2d_yaml, "");
+  }
+
+  template <typename SrvType>
+  decltype(auto) doSrvRequest(
+    const std::string & service, std::shared_ptr<typename SrvType::Request> request,
+    std::chrono::milliseconds timeout = default_service_timeout)
+  {
+    auto client = test_node_->create_client<SrvType>(service);
+    if (!client->wait_for_service(timeout)) {
+      std::runtime_error("The service '" + service + "' is not ready");
+    }
+
+    auto future = client->async_send_request(request);
+
+    if (
+      executor_.spin_until_future_complete(future, timeout) != rclcpp::FutureReturnCode::SUCCESS) {
+      RCLCPP_ERROR_STREAM(rclcpp::get_logger("rclcpp"), "Failed to call service " << service);
+      throw std::runtime_error(
+        "Failed to invoke service '" + service + "'. No response within timeout.");
+    }
+
+    return future.get();
+  }
+
+  template <typename SrvType>
+  decltype(auto) doEmptySrvRequest(
+    const std::string & service, std::chrono::milliseconds timeout = default_service_timeout)
+  {
+    auto request = std::make_shared<typename SrvType::Request>();
+    return doSrvRequest<SrvType>(service, request, timeout);
+  }
+
+  decltype(auto) doStdSrvsTriggerRequest(
+    const std::string & service, std::chrono::milliseconds timeout = default_service_timeout)
+  {
+    return doEmptySrvRequest<std_srvs::srv::Trigger>(service, timeout);
+  }
+
   template <typename Type>
   class SubscriptionWrapper
   {
   public:
-    static SubscriptionWrapper<Type> make(ros::NodeHandle & nh, const std::string & name)
+    static SubscriptionWrapper<Type> make(rclcpp::Node::SharedPtr node, const std::string & topic)
     {
       auto w = SubscriptionWrapper<Type>();
-      boost::function<void(const boost::shared_ptr<const Type> &)> cb =
-        [impl = w.impl_.get()](const auto & v) mutable {
-          impl->num_messages_++;
-          impl->last_message_ = *v;
-        };
-      w.impl_->subscriber_ = nh.subscribe<Type>(name, 1, cb);
+      std::function<void(const std::shared_ptr<const Type> &)> cb = [&](auto msg) {
```

Comment:
The `[&]` capture here looks suspicious, the `w` object will go out of scope after returning from the function. Perhaps store the reference to the impl itself?

Reply:
138913e9a424e4669c3685b47d7a3ce3bb239c92